wicketd/src/transceivers.rs

```diff
@@ -330,53 +332,63 @@ async fn fetch_transceiver_state(
-    modules = ModuleId::from_index_iter(present).unwrap();
-
-    // Vendor details.
-    let all_vendor_info = controller.vendor_info(modules).await?;
-    modules = all_vendor_info.modules;
-
-    // Power information.
-    let all_power = controller.power(modules).await?;
-    modules = all_vendor_info.modules;
-
-    // Datapath state.
-    let all_datapaths = controller.datapath(modules).await?;
-    modules = all_datapaths.modules;
-
-    // And monitors.
-    let all_monitors = controller.monitors(modules).await?;
-    modules = all_monitors.modules;
+    let present_modules = ModuleId::from_index_iter(
+        all_status
+            .iter()
+            .filter(|(_, st)| st.contains(ExtendedStatus::PRESENT))
+            .map(|(p, _st)| p),
+    )
+    .unwrap();
+
+    // Collect all available data.
+    let all_vendor_info = controller.vendor_info(present_modules).await?;
+    let all_power = controller.power(present_modules).await?;
+    let all_datapaths = controller.datapath(present_modules).await?;
+    let all_monitors = controller.monitors(present_modules).await?;
 
     // Now, combine everything.
     //
-    // If we failed any operation for a module, we'll insert `None` so that we
-    // still have whatever data we _could_ collect.
-    let mut out = Vec::with_capacity(modules.selected_transceiver_count());
-    for i in modules.to_indices() {
+    // For each operation, we'll record either the successful data or the error.
+    // This could lead to a lot of duplicated error messages, but at this point
+    // we'd rather be explicit, and it's possible for errors to be different.
+    let mut out =
+        Vec::with_capacity(present_modules.selected_transceiver_count());
+    for i in present_modules.to_indices() {
         let tr = Transceiver {
             port: format!("qsfp{i}"),
-            power: all_power.nth(i).copied(),
-            vendor: all_vendor_info.nth(i).cloned(),
-            status: all_status.nth(i).copied(),
-            datapath: all_datapaths.nth(i).cloned(),
-            monitors: all_monitors.nth(i).cloned(),
+            power: result_copied(&all_power, i),
+            vendor: result_cloned(&all_vendor_info, i),
+            status: result_copied(&all_status, i),
+            datapath: result_cloned(&all_datapaths, i),
+            monitors: result_cloned(&all_monitors, i),
         };
         out.push(tr);
     }
 
     Ok(out)
 }
+
+fn result_copied<T: Copy>(res: &ModuleResult<T>, i: u8) -> Result<T, String> {
+    res.nth(i).copied().ok_or_else(|| {
+        res.failures
+            .nth(i)
+            .map(|e| e.to_string())
+            .unwrap_or_else(|| String::from("Unknown failure"))
+    })
+}
+
+fn result_cloned<T: Clone>(res: &ModuleResult<T>, i: u8) -> Result<T, String> {
+    res.nth(i).cloned().ok_or_else(|| {
+        res.failures
+            .nth(i)
+            .map(|e| e.to_string())
+            .unwrap_or_else(|| String::from("Unknown failure"))
+    })
```

Comment:
Is "unknown failure" right here? When would this happen?

Reply:
This would be a violation of the invariant of the result type, meaning that the `i`th item neither succeeded nor failed. The message isn't great, but we're in uncharted territory so I don't have a better one.